keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+
+For features that fundamentally influence the workings of the entire cluster we need a mechanism that ensures that all members agree on the feature enablement. To address these challenges, we propose introducing a cluster-wide feature enablement mechanism that ensures consistent feature activation across all members of an etcd cluster.
+
+### Goals
+
+* Cluster-Wide Feature Consistency: Establish a server-side mechanism to ensure that specific features are either enabled or disabled consistently across all members of an etcd cluster.
+* Safe Feature Dependency: Provide a client-side mechanism that allows external systems and applications to reliably determine if a cluster-wide feature is enabled, enabling them to safely depend on and utilize that feature.
+
+### Non-Goals
+
+* Backporting a feature gate to a patch version is not supported.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+A developer is adding a new feature of a new write type. Because it requires a consensus write, the request should be rejected if the feature is not enabled or the request can be processed on all cluster members. 
+Today, each change in etcd API, requires development spenning at least 2 minor versions. First a raft proto change needs to be added and wait a release before it can be used. Only a release later it can be actually used to prevent a incompatible proto being used during upgrade/downgrade process. 
+
+With cluster level feature gate, a server would know if a feature is enabled on all members, and decide whether or not to accept the new write request.
+
+#### Story 2
+
+A Kubernetes developer would like use a new great etcd feature, but it is not available on all supported etcd versions. What can they do?
+Historically, Kubernetes avoided using any features that were not available on all supported versions of etcd.
+From K8s 1.31, Kubernetes added very complicated and fragile system: [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to detect etcd version, parse it and guess whether this feature based on a set of hardcoded etcd version. It does not really know if the feature is enabled by the feature `--experimental` flag.
+
+With a client-side mechanism to reliably determine if a cluster-wide feature is enabled,  would allow Kubernetes to immediately utilize new features.
+
+#### Story 3
+
+In a HA cluster, users would like to turn off an enabled feature. They need to restart each server node one by one with feature changed from enabled to disabled. Today, after the restarting process begins and before all nodes are restarted, the behavior of the feature is really undefined: it is enabled on some nodes and disabled on the other nodes.
```

Comment:
Would be good to note that it's impossible for etcd contributors to ensure that feature maintains correctness during the undefined state. Automatic testing would not be able to cover all possible edge cases in distributed systems, it's even too much for robustness tests.

Reply:
added.